text/0000-rand-crypto.md

```diff
@@ -0,0 +1,190 @@
+
+`rand::Rng` is not the subject of this RFC, but it is assumed here that this
+trait will have an API like the following:
+
+```rust
+pub trait Rng {
+    fn next_u32(&mut self) -> u32;
+    fn next_u64(&mut self) -> u64 { /* default impl here */ }
+    // possibly also next_u128
+    
+    // possibly also the following (no Result unlike try_fill above):
+    fn fill(&mut self, dest: &mut [u8]) { /* default impl here */ }
+}
+```
+
+
+# Motivation
+[motivation]: #motivation
+
+The [Rand revision RFC]
+brings up roughly this design several times:
+
+*   https://github.com/rust-lang/rfcs/pull/2106#issuecomment-322329159
+*   https://github.com/rust-lang/rfcs/pull/2106#issuecomment-323320722
+*   https://github.com/rust-lang/rfcs/pull/2106#pullrequestreview-56252714
+*   https://github.com/rust-lang/rfcs/pull/2106#issuecomment-323326062
+
+The appeal of this split is that cryptographic applications can use exactly the
+parts of `rand` they need, and crypto-algs have a very minimal trait filling a
+buffer and returning a result.
+
+Currently the [ring crate](https://github.com/briansmith/ring) implements its
+own version of [rand](https://github.com/rust-lang-nursery/rand)'s `Rng` and
+`OsRng` since it desires a slightly different interface. With this change, the
+"OS RNG" code need not be duplicated. Further, cryptographic PRNG
+implementations can be used by cryptographic-specific code as well as used by
+numerical applications through the implementation of `rand::Rng` for `CryptoRng`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Random number generation is a complex topic, and usage can be categorised into
+two main groups: cryptographic applications requiring secure *keys*, and
+non-cryptographic applications (which can be further divided into at least
+randomised algorithms, stochastic simulations, and games). For cryptographic
+applications, we recommend using a well-reviewed library; these libraries may
+or may not choose to use `crypto-rand` and `crypto-rng-os` to generate their
+random numbers. For non-cryptographic applications, users may make use of the
+`rand` crate [etc.]
+
+[Note that the `cryto-rand` and `crypto-rng-os` crates are only intended to be
+used directly by advanced cryptograhic users, hence the guide need not be very
+detailed on the subject, aside from a short preamble and link to the `rand`
+crate.]
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `CryptoRng` trait and `Error` enum are specified above. We should make it
+clear that although `Error::Unspecified` is directly accessible, `Error` could
+be extended in the future (is formal specification of this possible?).
+
+The `crypto-rng-os` crate should contain a single public member:
+
+```rust
+pub struct OsRng {
+    // private internals
+}
+
+impl OsRng {
+    pub fn new() -> OsRng {
```

Comment:
Would it be possible to move the fallibility into the constructor?  So the trait method would be infallible, but you'd use `OsRng::try_new()` to get an instance, which might fail with something like `Err(Error::NotYetSecurelySeeded)` or `Err(Error::NotAvailableOnPlatform)`.  (My recollection from the previous threads and sponge papers is that once the CSPRNG has sufficient entropy to start producing values, it will no longer fail.)